pkg/distributor/distributor.go

```diff
@@ -340,129 +376,188 @@ func (d *Distributor) Push(ctx context.Context, req *client.WriteRequest) (*clie
+
+			// Generate the sharding token based on the series labels without the HA replica
+			// label and dropped labels (if any)
+			key, err := d.tokenForLabels(userID, ts.Labels)
+			if err != nil {
+				return nil, err
+			}
 
-		validatedSeries, err := d.validateSeries(ts, userID)
+			validatedSeries, err := d.validateSeries(ts, userID)
 
-		// Errors in validation are considered non-fatal, as one series in a request may contain
-		// invalid data but all the remaining series could be perfectly valid.
-		if err != nil {
-			lastPartialErr = err
-		}
+			// Errors in validation are considered non-fatal, as one series in a request may contain
+			// invalid data but all the remaining series could be perfectly valid.
+			if err != nil {
+				lastPartialErr = err
+			}
 
-		// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
-		if validatedSeries == emptyPreallocSeries {
-			continue
+			// validateSeries would have returned an emptyPreallocSeries if there were no valid samples.
+			if validatedSeries == emptyPreallocSeries {
+				continue
+			}
+
+			seriesKeys = append(seriesKeys, key)
+			validatedTimeseries = append(validatedTimeseries, validatedSeries)
+			validatedSamples += len(ts.Samples)
 		}
+	}
 
-		keys = append(keys, key)
-		validatedTimeseries = append(validatedTimeseries, validatedSeries)
-		validatedSamples += len(ts.Samples)
+	if len(req.Metadata) > 0 {
+		for _, m := range req.Metadata {
+			key := d.tokenForMetadata(userID, m.MetricName)
+
+			err := d.validateMetadata(m.MetricMetadata, userID)
+
+			if err != nil {
+				// TODO: save the error to a partial error?
+				continue
+			}
+
+			metadataKeys = append(metadataKeys, key)
+			validatedMetadata = append(validatedMetadata, m)
+		}
 	}
+
 	receivedSamples.WithLabelValues(userID).Add(float64(validatedSamples))
+	receivedMetadata.WithLabelValues(userID).Add(float64(len(validatedMetadata)))
+
+	var subRing ring.ReadRing
+	subRing = d.ingestersRing
+
+	// Obtain a subring if required
+	if size := d.limits.SubringSize(userID); size > 0 {
+		h := client.HashAdd32(client.HashNew32(), userID)
+		subRing, err = d.ingestersRing.Subring(h, size)
+		if err != nil {
+			return nil, httpgrpc.Errorf(http.StatusInternalServerError, "unable to create subring: %v", err)
+		}
+	}
 
-	if len(keys) == 0 {
-		// Ensure the request slice is reused if there's no series passing the validation.
-		client.ReuseSlice(req.Timeseries)
+	if len(seriesKeys) == 0 && len(metadataKeys) == 0 {
+		// Ensure the request slice is reused if there's no series or metadata passing the validation.
+		client.ReuseSlice(req.Timeseries, req.Metadata)
 
 		return &client.WriteResponse{}, lastPartialErr
 	}
 
 	now := time.Now()
-	if !d.ingestionRateLimiter.AllowN(now, userID, validatedSamples) {
+	//TODO: might be a problem for customers over the tipping point. Maybe increase limits by around 5 - 10%?
+	totalN := validatedSamples + len(validatedMetadata)
```

Comment:
For rate-limiting, I have done the simplest thing we can do and unified the amounts. This might be unfeasible for tenants that are near the tipping point. But, having two separate limits brings its own set of complexities (e.g. do we drop samples if we're over the metadata limit?) and I would rather have a conversation instead of going with an implementation. 

Metadata as a whole will always be a subset of the data we receive, and the [current default is 100](https://github.com/prometheus/prometheus/blob/9d9bc524e5e893287cfeeab9995252f54b0624a8/config/config.go#L112) samples per send.

Given, metadata is sent once every minute by default and as a set (for now), the [bandwidth](https://docs.google.com/document/d/1LoCWPAIIbGSq59NG3ZYyvkeNb8Ymz28PUKbg_yhAzvE/edit#heading=h.pzpbsen14sdg) compared to samples is almost negligent. I don't think this could be a problem right now but it is certainly coupled to the implementation of remote write? But feel free to prove me wrong on this one.